server/src/main/java/org/elasticsearch/action/admin/indices/alias/TransportIndicesAliasesAction.java

```diff
@@ -222,25 +238,33 @@ protected void masterOperation(
                         }
                         break;
                     case REMOVE:
-                        for (String alias : concreteAliases(action, state.metadata(), index.getName())) {
+                        String[] matchingAliases = concreteAliases(action, state.metadata(), index.getName());
+                        AliasActionResult newResult = matchingAliases.length == 0
+                            ? AliasActionResult.REMOVE_MISSING
+                            : AliasActionResult.buildSuccess(action);
+                        actionResult = AliasActionResult.merge(actionResult, newResult);
+                        for (String alias : matchingAliases) {
                             finalActions.add(new AliasAction.Remove(index.getName(), alias, action.mustExist()));
                         }
                         break;
                     case REMOVE_INDEX:
+                        actionResult = AliasActionResult.buildSuccess(action);
                         finalActions.add(new AliasAction.RemoveIndex(index.getName()));
                         break;
                     default:
                         throw new IllegalArgumentException("Unsupported action [" + action.actionType() + "]");
                 }
             }
+            actionResults.add(actionResult);
```

Comment:
It took me a few minutes to convince myself that `actionResult` can never be null here, so I thought I'd write down why in order to possibly save someone else some time.
It can't because `concreteIndices` is never empty because we would have already had an exception by this point if it had been because IndicesAliasesRequest's indicesOptions always sets `allowNoIndices` to `false`.

Reply:
Thanks for reasoning this out. This code is pretty confusing, and would be easy to introduce a bug (assuming there's not one already!).
I just refactored this loop in a way that I think is easier to believe is bug free.